elks/tools/elf2elks/elf2elks.c

```diff
@@ -527,6 +531,60 @@ input_for_header (void)
 	"%" PRIu32 " data reloc(s).", text_n_rels, ftext_n_rels, data_n_rels);
 }
 
+static char symtab_filename[256];
+
+static void
+create_symtab (void)
+{
+    FILE *infp;
+    FILE *outfp;
+    char command[256];
+    char buf[128];
+
+    if (!symtab) return;
+
+    snprintf(symtab_filename, sizeof(symtab_filename), "%s-symXXXXXX", file_name);
+    mktemp(symtab_filename);
+    outfp = fopen(symtab_filename, "w");
+    if (!outfp)
+        error("Can't create temp file %s\n", symtab_filename);
+    snprintf(command, sizeof(command),
+        "ia16-elf-nm %s | sed -e '/&/d; /!/d' | sort", file_name);
+    infp = popen(command, "r");
+    if (!infp)
+        error("Can't popen '%s'\n", command);
+    int lastwastext = 0;
+    while(fgets(buf, sizeof(buf), infp) != NULL) {
+        int addr;
+        unsigned char type;
+        char name[128];
+
+        /* 0x10000+ is text, 0x20000+ is far text 0x30000+ is data */
+        if (sscanf(buf, "%x %c %s", &addr, &type, name) == 3) {
```

Comment:
The hard-coded linear addresses are kind of an implementation detail in the linker script.  I think it would be better to use the addresses that are actually provided in the intermediate ELF file &mdash; namely, `text_sh->sh_addr`, `ftext_sh->sh_addr`, and `data_sh->sh_addr` &mdash; similar to what `elf2elks` already does.

The `->sh_addr` values should already be available at this point in time, since `input_for_header ()` has already been called.

Reply:
Ok. It was a little tricky, given that there may be cases where data_sh and/or bss_sh could be NULL. Does the arithmetic look correct for calculating the data_start and data_end values?

Reply:
Hello @ghaerr,

Thanks &mdash; looks OK for now.  On second thought, probably a better way to gauge the segment that a symbol _foo_ belongs to, would be to consult its corresponding _foo_`!` symbol.  But this will probably require maintaining more complex data structures (or perhaps an `awk` script).  Perhaps this part is best left as an exercise for the future. :no_mouth: 

Thank you!